source/common/buffer/watermark_buffer.cc

```diff
@@ -63,12 +63,6 @@ uint64_t WatermarkBuffer::reserve(uint64_t length, RawSlice* iovecs, uint64_t nu
   return bytes_reserved;
 }
 
-Api::IoCallUint64Result WatermarkBuffer::write(Network::IoHandle& io_handle) {
-  Api::IoCallUint64Result result = OwnedImpl::write(io_handle);
-  checkLowWatermark();
```

Comment:
Just to understand better where/how do we check for `LowWatermark` in the new implementation?

The reason that I am asking is that the watermark tests are the ones that are failing.

Reply:
In WatermarkBuffer::drain(). Actually, the change short circuits the write by just passing the buffer to the io handle. The test that fails in ci is //test/common/network:connection_impl_test. Is it failing in a different place in your testbed?

Reply:
In my dev box the tests that are failing are:
1.  IpVersions/ConnectionImplTest.WriteWithWatermarks/IPv4
2.  IpVersions/ConnectionImplTest.WriteWithWatermarks/IPv6
3. IpVersions/ConnectionImplTest.WatermarkFuzzing/IPv4 (timesout)
4. IpVersions/ConnectionImplTest.WatermarkFuzzing/IPv6 (timesout)

Reply:
Thanks @davinci26 for the offline help with this! The issue seems to be a combination of level triggered epoll with a delayed cleanup of the os_sys_calls mock in the WriteWithWatermarks tests.